modules/sdk/src/main/scala/org/tessellation/sdk/domain/healthcheck/consensus/HealthCheckConsensusRound.scala

```diff
@@ -0,0 +1,122 @@
+package org.tessellation.sdk.domain.healthcheck.consensus
+
+import cats.Applicative
+import cats.effect.{Clock, Ref, Spawn}
+import cats.syntax.contravariantSemigroupal._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.gossip.Gossip
+import org.tessellation.sdk.domain.healthcheck.consensus.types._
+
+class HealthCheckConsensusRound[F[_]: Spawn: Clock, A <: HealthCheckStatus, B <: ConsensusHealthStatus[A]](
+  key: HealthCheckKey,
+  roundId: HealthCheckRoundId,
+  driver: HealthCheckConsensusDriver[A, B],
+  startedAt: Long,
+  peers: Ref[F, Set[PeerId]],
+  roundIds: Ref[F, Set[HealthCheckRoundId]],
+  proposals: Ref[F, Map[PeerId, B]],
+  gossip: Gossip[F],
+  selfId: PeerId
+) {
+
+  def start: F[Unit] =
+    Spawn[F].start {
+      sendProposal
+    }.void
+
+  def getPeers: F[Set[PeerId]] = peers.get
+
+  def manageAbsent(absentPeers: Set[PeerId]): F[Unit] =
+    proposals.get
+      .map(_.keySet)
+      .map(absentPeers -- _)
+      .flatMap { toRemove =>
+        peers.modify { p =>
+          (p -- toRemove, ())
+        }
+      }
+
+  def isFinished: F[Boolean] = allProposalsReceived
+
+  def processProposal(proposal: B): F[Unit] =
+    if (proposal.key == key) {
+      proposals.modify { m =>
+        def updated =
+          m.get(proposal.owner)
+            .fold(m + (proposal.owner -> proposal))(_ => m)
+
+        (updated, ())
+      }.flatMap { _ =>
+        roundIds.modify { m =>
+          (m + proposal.roundId, ())
+        }
+      }
+    } else Applicative[F].unit
+
+  def getRoundIds: F[Set[HealthCheckRoundId]] = roundIds.get
+
+  def addParallelRounds(key: HealthCheckKey)(roundIds: Set[HealthCheckRoundId]): F[Unit] = ???
+
+  def calculateOutcome: F[HealthCheckConsensusDecision] = {
+    def status: A = ???
+
+    proposals.get.map(_.values.toList).map { received =>
+      driver.calculateConsensusOutcome(key, status, selfId, received)
+    }
+  }
+
+  def manage: F[Unit] =
+    Clock[F].monotonic.map(_.toNanos).flatMap { currentTime =>
```

Comment:
Not sure that conversion to nanos is necessary. Depends on how it's going to be used probably. But you can compare `FiniteDuration`s with each other with e.g. `< > ==` and also subtract and add with success I think, so not sure if you need to switch to `Long` here.

Reply:
You'd want to compare a point in time to check if some delta time has elapsed. 
see: https://github.com/typelevel/fs2/blob/main/core/shared/src/main/scala/fs2/Stream.scala#L3166